aidial_adapter_openai/embeddings/azure_ai_vision.py

```diff
@@ -0,0 +1,153 @@
+from aidial_adapter_openai.utils.resource import Resource
+
+# The latest Image Analysis API offers two models:
+# * version 2023-04-15 which supports text search in many languages,
+# * the legacy 2022-04-11 model which supports only English.
+_VERSION_PARAMS = {
+    "api-version": "2024-02-01",
+    "model-version": "2023-04-15",
+}
+
+
+def _get_auth_headers(creds: OpenAICreds) -> dict[str, str]:
+    if "api_key" in creds:
+        return {"Ocp-Apim-Subscription-Key": creds["api_key"]}
+
+    if "azure_ad_token" in creds:
+        return {"Authorization": f"Bearer {creds['azure_ad_token']}"}
+
+    raise ValueError("Invalid credentials")
+
+
+class VectorizeResponse(BaseModel):
+    class Config:
+        extra = "allow"
+
+    vector: List[float]
+
+
+async def embeddings(
+    creds: OpenAICreds,
+    deployment: str,
+    endpoint: str,
+    file_storage: FileStorage | None,
+    data: dict,
+) -> EmbeddingResponse:
+    input = EmbeddingsRequest.parse_obj(data)
+
+    async def on_text(text: str) -> str:
+        return text
+
+    async def on_attachment(attachment: Attachment) -> Resource:
+        return await AttachmentResource(attachment=attachment).download(
+            file_storage
+        )
+
+    inputs_iter: AsyncIterator[str | Resource] = collect_embedding_inputs(
+        input,
+        on_text=on_text,
+        on_attachment=on_attachment,
+    )
+
+    inputs: List[str | Resource] = [input async for input in inputs_iter]
+
+    async def _get_embedding(input: str | Resource) -> VectorizeResponse:
+        if isinstance(input, str):
+            return await _get_text_embedding(creds, endpoint, input)
+        elif isinstance(input, Resource):
+            return await _get_image_embedding(creds, endpoint, input)
+        else:
+            assert_never(input)
+
+    tasks: List[asyncio.Task[VectorizeResponse]] = [
+        asyncio.create_task(_get_embedding(input)) for input in inputs
+    ]
+
+    responses = await asyncio.gather(*tasks)
+    vectors = [
+        Embedding(embedding=r.vector, index=idx)
+        for idx, r in enumerate(responses)
+    ]
+
+    n = len(vectors)
+    usage = Usage(prompt_tokens=n, total_tokens=n)
+
+    return EmbeddingResponse(model=deployment, data=vectors, usage=usage)
+
+
+async def _get_image_embedding(
+    creds: OpenAICreds, endpoint: str, resource: Resource
+) -> VectorizeResponse:
+    resp = await get_http_client().post(
+        url=endpoint.rstrip("/") + "/computervision/retrieval:vectorizeImage",
+        content=resource.data,
+        headers={
+            **_get_auth_headers(creds),
+            "content-type": resource.type,
+        },
+        params=_VERSION_PARAMS,
+    )
+
+    resp.raise_for_status()
+    return VectorizeResponse.parse_obj(resp.json())
+
+
+async def _get_text_embedding(
+    creds: OpenAICreds, endpoint: str, text: str
+) -> VectorizeResponse:
+    resp = await get_http_client().post(
```

Comment:
Currently we have both `aiohttp` (for `dalle3` and `gpt 4 multi modal`) and `httpx` as http clients 

Also, look like `httpx` still have performance issue owing to `anyio.Lock` :(
https://github.com/encode/httpx/issues/3215

I think we can use aiohttp session here for that reason